### ISSUE TITLE
ERC721Votes._afterTokenTransfer assumes 1 token = 1 vote

### ISSUE THREAD

**Issue by RitzyDevBox** [OpenZeppelin/openzeppelin-contracts#3997]
I found this while unit testing my contract.  I wanted to providing more voting power per nft, mostly because I've written a governor which handles partial voting.

I overriden the `_getVotingUnits` like so:

    function _getVotingUnits(address account) internal view virtual override returns (uint256) {
        return balanceOf(account) * 10000;
    }

When minting and delegating I'm getting the wrong result:

    await nft.mint(owner.address, { value: ethers.utils.parseUnits("1", 18)})
    await nft.delegate(owner.address);

    ownerVotes = await nft.getVotes(owner.address);
    console.log(`owner votes: ${ownerVotes.toString()}`);

    await nft.mint(owner.address, { value: ethers.utils.parseUnits("1", 18)})
    await nft.delegate(owner.address);

    ownerVotes = await nft.getVotes(owner.address);
    console.log(`owner votes: ${ownerVotes.toString()}`);

this outputs 

     10000
     10001

It increments sequentially after.  The issue is that `_transferVotingUnits(from, to, batchSize )` does not take the units into account.

    
 2. `_afterTokenTransfer` does not apply the Units to the batch size

`abstract contract ERC721Votes is ERC721, Votes {
     internal uint256 votingPowerBase = 1

    _setVotingPowerUnits(uint256 _votingPowerBase) internal virtual {
        votingPowerBase = _votingPowerBase
    }
    
    getVotingPowerUnits(uint256 _votingPowerBase) public virtual returns (uint 256) {
        return votingPower;
    }

    /**
     * @dev See {ERC721-_afterTokenTransfer}. Adjusts votes when tokens are transferred.
     *
     * Emits a {IVotes-DelegateVotesChanged} event.
     */
    function _afterTokenTransfer(
        address from,
        address to,
        uint256 firstTokenId,
        uint256 batchSize
    ) internal virtual override {
        _transferVotingUnits(from, to, batchSize * VotingPowerBase );
        super._afterTokenTransfer(from, to, firstTokenId, batchSize);
    }

    /**
     * @dev Returns the balance of `account`.
     */
    function _getVotingUnits(address account) internal view virtual override returns (uint256) {
        return balanceOf(account) * VotingPowerBase ;
    }`

I'm not sure how to increment the units otherwise, seeing as even if I override this directly in my contract it will still end up calling the base.

**Comment by Amxx** [comment]
Hello @RitzyDevBox 

The `_getVotingUnits` function is used to know how much weight to move when an accounts change its delegation.
If Alice is previously delegated to Bob, but is changing her delegate to Charles, then we remove `_getVotingUnits(Alice)` from Bob and add that to Charles.

Indeed, it is not accounted for when doing a transfer.

Your rewrite of `ERC721Votes` looks good, and IMO you rewritting it is probably the best thing to do. 
An alternative would be to write you own Governor module that does

```
    function _getVotes(
        address account,
        uint256 blockNumber,
        bytes memory /*params*/
    ) internal view virtual override returns (uint256) {
        return token.getPastVotes(account, blockNumber) * VotingPowerBase;
    }
```

---

We've had a few request for custom weight per token ... which is "easy" to do for the transfer part, but unresolved for the "re-delegate" part.

---

If (like in your case) all token have the same weight, then we could had a new function 
```
_votesPerToken() internal view virtual returns (uint256) { return 1; }
```
that devs could override.

Do you want to submit a PR for that?

**Comment by frangio** [comment]
`_votesPerToken` should be `pure`, not `view`, to make sure that the value is not dynamic.

**Comment by Amxx** [comment]
I commented on #4003.

TLDR: I think that should all be addressed in the governor by doing `return token.getPastVotes(account, blockNumber) * VotingPowerBase;`

**Comment by frangio** [comment]
Why do you need to scale it by a uniform amount? Seems like it doesn't make a difference.

**Comment by frangio** [comment]
> The issue is then your governor is tracking voting power and not your NFT

I don't really agree with this. The NFT continues tracking voting power. The governor simply scales it up, or you could say it uses it with different granularity.

**Comment by frangio** [comment]
Based on comments by @Amxx which I agree with, It doesn't seem we're open to this proposal. Personally I would be in favor of adding something like we're discussing in https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3996.in ERC721.
